WIP - apiextensions: low-level converter wiring – towards versioning

sttts opened the pull request:
- add crdConversionRESTOptionsGetter to wire in custom conversions in storage path
- wire through custom conversions in serializers for request de- and encoding.

mbohlool committed:
SQUASH: fallbackScheme

sttts committed:
apimachinery: add UnstructuredList support to versioning codec

sttts committed:
apiextensions: wire custom converter into storage and request codecs

sttts commented:
/cc @mbohlool

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:
```diff
@@ -170,17 +171,20 @@ func (c *codec) Encode(obj runtime.Object, w io.Writer) error {
-		}
-		if targetGVK == objGVK {
-			return c.encoder.Encode(obj, w)
+		// Only avoid conversion if unstructured is not a list, giving the converter a chance to convert list items.
```
When did you find a list that asserted the "correct" group/version, but actually contained types in the wrong group/version?

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_test.go:
```diff
@@ -59,7 +59,7 @@ func newStorage(t *testing.T) (customresource.CustomResourceStorage, *etcdtestin
 	)
 
-	typer := apiserver.UnstructuredObjectTyper{
+	typer := apiserver.unstructuredObjectTyper{
```
This doesn't compile, does it?

sttts replied:
No, it doesn't. We shouldn't leak from the apiserver into the registry in the first place though.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -640,3 +683,29 @@ func (in crdStorageMap) clone() crdStorageMap {
+			ret.StorageConfig.Codec,
+			t.converter,
+			&unstructuredCreator{},
+			discovery.NewUnstructuredObjectTyper(nil),
```
reminder to myself, I promised to move this to the extensions apiserver because it's the only consumer and it's misnamed.

sttts replied:
We might want to move all of these into a scheme package here.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -640,3 +683,29 @@ func (in crdStorageMap) clone() crdStorageMap {
+// and custom encoder and decoder version.
+type crdConversionRESTOptionsGetter struct {
+	delegate       generic.RESTOptionsGetter
+	converter      runtime.ObjectConvertor
```
please doc granularity.  All CRDs or one group? or something else?

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -535,9 +548,37 @@ func (c crdObjectConverter) ConvertFieldLabel(version, kind, label, value string
+		in.GetObjectKind().SetGroupVersionKind(gvk)
+
+		for i := range unstructList.Items {
+			out, err := c.unstructuredDelegate.ConvertToVersion(&unstructList.Items[i], target)
```
Seems weird to directly delegate instead of recursing.

sttts replied:
We can't recurse because the underlying converter will be different (think of declarative conversion some day). Hence, we make this composition, so the generic crd converter can call the delegate converter doing the actual work on non-lists.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -517,10 +522,18 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
-	unstructured.UnstructuredObjectConverter
-	clusterScoped bool
+	unstructuredDelegate unstructured.UnstructuredObjectConverter
+	clusterScoped        bool
```
cruft?

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -382,12 +382,17 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
-		UnstructuredTyper: discovery.NewUnstructuredObjectTyper(),
+	typer := unstructuredObjectTyper{
+		delegate:          parameterScheme,
+		unstructuredTyper: discovery.NewUnstructuredObjectTyper(),
```
This is on its way to being a lot more opinionated.  How about starting a `scheme` package or some such?

deads2k commented:
I think I see where you're going with it.  It makes sense.

Can you summarized the intent in the description?  It appears to me that you're working towards building a mutable object converter that is going to be used inside of the "normal" codec flow and will thus allow you to take a lock and change the converter as needed.

deads2k commented:
@kubernetes/sig-api-machinery-misc @lavalamp

lavalamp commented:
/assign @mbohlool

lavalamp commented:
Can this be rebased on top of #63830? Or is it trying to be an alternative approach?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -535,9 +548,37 @@ func (c crdObjectConverter) ConvertFieldLabel(version, kind, label, value string
+func (c crdObjectConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
+	kind := in.GetObjectKind().GroupVersionKind()
+	if kind.Empty() {
+		return in, nil
```
here the analysis and a possible fix: https://github.com/mbohlool/kubernetes/pull/5

deads2k commented:
> Can this be rebased on top of #63830? Or is it trying to be an alternative approach?

My read is that this is doing the underlying wiring that is logically needed in #63830 and @sttts is helping actually do it instead of simply pushing a many comments into the pull.

sttts commented:
> My read is that this is doing the underlying wiring that is logically needed in #63830

Yes, this is 90% the same code as in #63830, minus adding the multiple versions. The goal was to iterate quicker with reviews than mixing lower level plumbing with API discussion in a big API implementing PR. The later takes a magnitude longer for reviewed and to get it merged. Also having some soak time for those low-level changes at the heart of CRDs wouldn't harm.

These changes here are also blocking work on pruning and defaulting, which puts it on the critical path towards code-freeze (and with pruning being prerequisite for GA also for promotion of CRDs).

I don't feel strongly about which PR we finally merge (code is mostly the same), as long as we get it in this week. So if we put enough priority on #63830 both for review and for addressing comments, that's fine as well.
